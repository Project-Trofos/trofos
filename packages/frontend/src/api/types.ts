export type Project = {
  id: number;
  pname: string;
  pkey: string | null;
  description: string | null;
  course_id: number | null;
  public: boolean;
  created_at: string;
  course?: Course;
  is_archive: boolean | null;
};

export type Course = {
  id: number;
  code: string;
  startYear: number;
  startSem: number;
  endYear: number;
  endSem: number;
  cname: string;
  description: string | null;
  public: boolean;
  created_at: string;
  shadow_course: boolean;
  is_archive: boolean | null;
};

export type User = {
  user_email: string;
  user_display_name: string;
  user_id: number;
  projects: Project[];
  courses: CourseRoles[];
  basicRoles: BasicRoles[];
};

export type BasicRoles = {
  user_email: string;
  role_id: number;
};

export type CourseRoles = {
  id: number;
  user_email: string;
  role_id: number;
  course_id: number;
};

export type CreateUserRequest = {
  userEmail: string;
  newPassword: string;
};

export type RegisterUser = {
  userEmail: string;
  newPassword: string;
  userDisplayName: string;
};

export type UserCourseRoleRequest = {
  id: number;
  userRole: number;
  userId: number;
};

export type Role = {
  role_name: string;
  id: number;
};

export type Action = {
  action: string;
};

export type ActionOnRole = {
  id: number;
  action: string;
};

export type ActionsOnRoles = {
  id: number;
  role_name: string;
  actions: Action[];
};

export type ProjectData = Project & {
  course: Course | null;
  users: UserData[];
  sprints: {
    id: number;
    name: string;
  }[];
  backlogStatuses: Omit<BacklogStatusData, 'project_id'>[];
  telegramChannelLink: string;
  owner_id: number | null;
};

export type UserData = {
  user: Pick<User, 'user_id' | 'user_email' | 'user_display_name' | 'courses'>;
};

export type ScrumBoardUserData = {
  user: {
    user_id: number | null;
    user_display_name: string;
    user_email: string;
  };
};

export type Milestone = {
  id: number;
  name: string;
  start_date: string;
  deadline: string;
  course_id: number;
  created_at: string;
};

export type Announcement = {
  id: number;
  title: string;
  content: string;
  user_id: number;
  created_at: string;
  updated_at?: string;
};

// CourseData type returned by BE api. We perform transformResponse to convert the data into type CourseData
export type CourseDataResponse = Omit<CourseData, 'users'> & {
  courseRoles: UserData[];
};

export type CourseData = Course & {
  milestones: Milestone[];
  announcements: Announcement[];
  users: UserData[];
};

export type UserOnRolesOnCourse = {
  id: number;
  user_id: number;
  role_id: number;
  course_id: number;
  role: {
    id: number;
    role_name: string;
  };
};

export type BacklogStatusData = {
  name: string;
  type: 'todo' | 'in_progress' | 'done';
  project_id: number;
  order: number;
};

export type Comment = {
  comment_id: number;
  backlog_id: number;
  project_id: number;
  commenter_id: number;
  content: string;
  created_at: string;
  updated_at: string | null;
  commenter: {
    created_at: string;
    project_id: string;
    user: {
      user_id: number;
      user_email: string;
      user_display_name: string;
    };
  };
};

export type CommentFieldsType = {
  projectId: number;
  backlogId: number;
  commenterId: number;
  content: string;
};

export enum BacklogStatus {
  TODO = 'To do',
  DONE = 'Done',
}

export type BacklogPriority = 'very_high' | 'high' | 'medium' | 'low' | 'very_low' | null;

export type Backlog = {
  backlog_id: number;
  summary: string;
  type: 'story' | 'task' | 'bug';
  priority: BacklogPriority;
  reporter_id: number;
  assignee_id: number | null;
  sprint_id: number | null;
  points: number | null;
  description: string | null;
  project_id: number;
  status: BacklogStatus | string;
  assignee: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
  epic?: Epic | null;
};

export type Epic = {
  epic_id: number;
  project_id: number;
  name: string;
  description: string | null;
};

export type DefaultBacklog = {
  backlog_id?: number;
  summary?: string;
  type?: 'story' | 'task' | 'bug';
  priority?: BacklogPriority;
  reporter_id?: number;
  assignee_id?: number | null;
  sprint_id?: number | null;
  points?: number | null;
  description?: string | null;
  project_id?: number;
  status?: BacklogStatus | string;
  assignee?: {
    created_at: string;
    project_id: number;
    user_id: number;
    user: {
      user_email: string;
      user_display_name: string;
    };
  } | null;
  epic_id?: number | null;
};

export type BacklogUpdatePayload = {
  projectId: number;
  backlogId: number;
  srcSprintId?: number | null;
  fieldToUpdate: Partial<Backlog>;
};

export enum BacklogHistoryType {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
}

export type BacklogHistory = {
  history_type: BacklogHistoryType;
  date: string;
} & Omit<Backlog, 'assignee' | 'summary' | 'description'>;

export type UpdateUserRolePayload = {
  userId: number;
  newRoleId: number;
};

export type ProjectGitLink = {
  project_id: number;
  repo: string;
};

export type ProjectGitLinkData = {
  projectId: number;
  repoLink: string;
};

export type ProjectUserSettings = {
  project_id: number;
  user_id: number;
  email_notification: boolean;
};

export type Settings = {
  current_year: number;
  current_sem: number;
};

export type RetrospectiveVote = {
  id: number;
  retro_id: number;
  user_id: number;
  type: RetrospectiveVoteType;
};

export type Retrospective = {
  id: number;
  sprint_id: number;
  content: string;
  type: RetrospectiveType;
  score: number;
  votes: RetrospectiveVote[];
  is_action_taken: boolean;
};

export enum RetrospectiveType {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  ACTION = 'action',
}

export enum RetrospectiveVoteType {
  UP = 'up',
  DOWN = 'down',
}

export type CourseImportCsvPayload = {
  courseId: number;
  payload: FormData;
};

export type Feedback = {
  id: number;
  user_id?: number;
  user: {
    user_email: string;
    user_display_name: string;
  };
  sprint_id: number;
  content: string;
  created_at: string;
  updated_at?: string;
};

export type OAuth2Payload = {
  code: string;
  state: string;
  callbackUrl: string;
};

export type Invite = {
  project_id: number;
  email: string;
  expiry_date: Date;
};

export type UserApiKey = {
  id: number;
  user_id: number;
  api_key: string;
  created_at: string;
  last_used: string | null;
  active: boolean;
};

export type UserGuideQueryResponse = {
  answer: string;
  links: Array<string>;
};

export type FeatureFlag = {
  feature_name: string;
  active: boolean;
};

export type ProjectAssignment = {
  id: number;
  targetProjectId: number;
  sourceProjectId: number;
};
