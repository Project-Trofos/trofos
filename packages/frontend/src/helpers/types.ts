export type PickRename<T, K extends keyof T, R extends PropertyKey> = Pick<Omit<T, K> & { [P in R]: T[K] }, R>;
