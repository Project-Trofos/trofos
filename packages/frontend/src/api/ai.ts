import trofosApiSlice from '.';
import { UserGuideQueryResponse, UserGuideRecommendation } from './types';

const extendedApi = trofosApiSlice.injectEndpoints({
  endpoints: (builder) => ({
    answerUserGuideQuery: builder.mutation<UserGuideQueryResponse, { query: string; isEnableMemory: boolean }>({
      query: ({ query, isEnableMemory }) => ({
        url: `ai/userGuideQuery`,
        credentials: 'include',
        method: 'POST',
        body: { query, isEnableMemory },
      }),
    }),
    recommendUserGuideSections: builder.mutation<UserGuideRecommendation[], void>({
      query: () => ({
        url: `ai/recommendUserGuide`,
        method: 'POST',
        credentials: 'include',
      }),
    }),
  }),
});

export const { useAnswerUserGuideQueryMutation, useRecommendUserGuideSectionsMutation } = extendedApi;
