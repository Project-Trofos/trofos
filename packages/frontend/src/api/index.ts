import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Define a service using a base URL and expected endpoints
const trofosApiSlice = createApi({
  reducerPath: 'trofosApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NODE_ENV === 'test' ? 'http://localhost:3001/api' : '/api',
    prepareHeaders: (headers) => {
      headers.set('Cache-control', 'no-cache');
      return headers;
    },
  }),
  tagTypes: [
    'Project',
    'UserInfo',
    'Course',
    'Backlog',
    'BacklogHistory',
    'User',
    'Sprint',
    'ActionsOnRoles',
    'CourseRoles',
    'ProjectRoles',
    'Settings',
    'StandUpHeader',
    'StandUp',
    'Retrospective',
    'ProjectUserSettings',
    'Feedback',
  ],
  endpoints: () => ({}),
});

// Define a service using a base URL and expected endpoints
export const nusmodsApiSlice = createApi({
  reducerPath: 'nusmodsApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.nusmods.com/v2/2022-2023' }),
  endpoints: () => ({}),
});

export default trofosApiSlice;
