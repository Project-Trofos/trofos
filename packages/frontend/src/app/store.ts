import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
