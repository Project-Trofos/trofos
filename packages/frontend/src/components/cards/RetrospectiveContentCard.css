.retrospective-content-card {
  margin: 4px 0;
}

.retrospective-content-card .ant-card-body {
  padding: 6px 18px;
}

.retrospective-content-body {
  display: flex;
}

.vote-container {
  min-width: 60px;
  margin-right: 15px;
  display: flex;
  align-items: center;
}

.vote-container button {
  border: none;
  padding: 0;
}

.vote-container .vote-count {
  padding: 0 10px;
}

.vote-container button.up.ant-btn:hover,
.vote-container button.up.is-selected {
  background-color: rgb(0 0 0 / 4%);
  color: #42a7ee;
}

.vote-container button.down.ant-btn:hover,
.vote-container button.down.is-selected {
  background-color: rgb(0 0 0 / 4%);
  color: #f34132;
}

.retrospective-content-text {
  display: flex;
  align-items: center;
  overflow-x: hidden;
  width: 100%;
}

.retrospective-content-delete-container {
  min-width: 30px;
  margin-left: 10px;
  display: flex;
  align-items: center;
}

.retrospective-content-delete-container-create {
  min-width: 80px;
  margin-left: 10px;
  display: flex;
  align-items: center;
}

.retrospective-content-delete-container button,
.retrospective-content-delete-container-create button {
  border: none;
  padding: 3px;
}

.retrospective-content-delete-container button.delete.ant-btn:hover,
.retrospective-content-delete-container-create button.delete.ant-btn:hover {
  background-color: rgb(0 0 0 / 4%);
  color: #f34132;
}
