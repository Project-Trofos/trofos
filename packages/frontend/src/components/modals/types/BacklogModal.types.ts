export interface BacklogFormFields extends FormData {
  projectId: number;
}

export type BacklogSelect = {
  id: string;
  name: string;
};
