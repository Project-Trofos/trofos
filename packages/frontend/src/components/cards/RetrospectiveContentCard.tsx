import React from 'react';
import { Button, Card, message } from 'antd';
import { ArrowDownOutlined, ArrowUpOutlined, DeleteOutlined } from '@ant-design/icons';
import { isEqual } from 'lodash';
import {
  useAddRetrospectiveVoteMutation,
  useDeleteRetrospectiveVoteMutation,
  useUpdateRetrospectiveVoteMutation,
  useDeleteRetrospectiveMutation,
} from '../../api/socket/retrospectiveHooks';
import { Retrospective, RetrospectiveType, RetrospectiveVoteType } from '../../api/types';
import './RetrospectiveContentCard.css';
import { getErrorMessage } from '../../helpers/error';
import { confirmDeleteRetrospective } from '../modals/confirm';
import BacklogCreationModal from '../modals/BacklogCreationModal';
import { useGetUserInfoQuery } from '../../api/auth';

function RetrospectiveContentCard(props: RetrospectiveContentCardProps): JSX.Element {
  const { retroEntry, readOnly } = props;
  const [addRetrospectiveVote] = useAddRetrospectiveVoteMutation();
  const [updateRetrospectiveVote] = useUpdateRetrospectiveVoteMutation();
  const [deleteRetrospectiveVote] = useDeleteRetrospectiveVoteMutation();
  const [deleteRetrospective] = useDeleteRetrospectiveMutation();
  const { data: userInfo } = useGetUserInfoQuery();

  const getCurrentVote = (): RetrospectiveVoteType => {
    return retroEntry.votes?.[0]?.type || null;
  };

  const currentVote = getCurrentVote();

  const handleVoteClick = async (vote: RetrospectiveVoteType) => {
    try {
      if (!currentVote) {
        await addRetrospectiveVote({
          retroId: retroEntry.id,
          type: vote,
          sprintId: retroEntry.sprint_id,
          retroType: retroEntry.type,
        }).unwrap();
      } else if (vote === currentVote) {
        await deleteRetrospectiveVote({
          retroId: retroEntry.id,
          sprintId: retroEntry.sprint_id,
          retroType: retroEntry.type,
        }).unwrap();
      } else {
        await updateRetrospectiveVote({
          retroId: retroEntry.id,
          type: vote,
          sprintId: retroEntry.sprint_id,
          retroType: retroEntry.type,
        }).unwrap();
      }
    } catch (err) {
      console.error(err);
    }
  };

  const handleDelete = async () => {
    try {
      confirmDeleteRetrospective(async () => {
        await deleteRetrospective({
          retroId: retroEntry.id,
          sprintId: retroEntry.sprint_id,
          retroType: retroEntry.type,
        }).unwrap();
      });
    } catch (err) {
      message.error(getErrorMessage(err));
    }
  };

  return (
    <Card className="retrospective-content-card">
      <div className="retrospective-content-body">
        <div className="vote-container">
          <Button
            onClick={() => handleVoteClick(RetrospectiveVoteType.UP)}
            className={`up ${currentVote === RetrospectiveVoteType.UP ? 'is-selected' : ''}`}
            disabled={readOnly}
          >
            <ArrowUpOutlined />
          </Button>
          <span className="vote-count">{retroEntry.score}</span>
          <Button
            onClick={() => handleVoteClick(RetrospectiveVoteType.DOWN)}
            className={`down ${currentVote === RetrospectiveVoteType.DOWN ? 'is-selected' : ''}`}
            disabled={readOnly}
          >
            <ArrowDownOutlined />
          </Button>
        </div>
        <div className="retrospective-content-text">{retroEntry.content}</div>
        <div
          className={`retrospective-content-delete-container${
            retroEntry && retroEntry.type == RetrospectiveType.ACTION ? '-create' : ''
          }`}
        >
          <Button onClick={() => handleDelete()} className={'delete'} disabled={readOnly}>
            <DeleteOutlined />
          </Button>
          {retroEntry && retroEntry.type == RetrospectiveType.ACTION ? (
            <BacklogCreationModal
              title={'create'}
              modalKey={retroEntry.id}
              defaultBacklog={{
                reporter_id: userInfo?.userId ?? undefined,
                summary: retroEntry.content ?? undefined,
                points: 2,
                type: 'task',
                priority: 'medium',
              }}
              disabled={retroEntry.is_action_taken}
              retrospective={retroEntry}
            />
          ) : (
            <></>
          )}
        </div>
      </div>
    </Card>
  );
}

type RetrospectiveContentCardProps = {
  retroEntry: Retrospective;
  readOnly?: boolean;
};

export default React.memo(RetrospectiveContentCard, (prevRetro, currRetro) => isEqual(prevRetro, currRetro));
