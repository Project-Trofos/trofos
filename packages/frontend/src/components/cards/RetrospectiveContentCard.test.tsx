import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import RetrospectiveContentCard from './RetrospectiveContentCard';
import store from '../../app/store';
import { Retrospective, RetrospectiveType } from '../../api/types';

describe('RetrospectiveContentCard test', () => {
  const setup = (retro: Retrospective) => {
    const { baseElement, debug } = render(
      <Provider store={store}>
        <RetrospectiveContentCard retroEntry={retro} />
      </Provider>,
    );

    return { baseElement, debug };
  };

  const MOCK_RETRO_ENTRY: Retrospective = {
    id: 1,
    sprint_id: 1,
    content: 'Retrospective text',
    type: RetrospectiveType.ACTION,
    score: 1,
    votes: [],
    is_action_taken: false,
  };

  it('renders retrospective card with correct details', () => {
    const { baseElement } = setup(MOCK_RETRO_ENTRY);
    expect(screen.getByText(MOCK_RETRO_ENTRY.content)).toBeInTheDocument();
    expect(screen.getByText(MOCK_RETRO_ENTRY.score)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });

  it('should have non-disabled create button if is_action_taken is false in action retrospective', () => {
    const { baseElement } = setup(MOCK_RETRO_ENTRY);
    screen.debug();
    // get create button
    const createButton = screen.getByRole('button', { name: /create/i });

    // Assert that the button is not disabled
    expect(createButton).not.toBeDisabled();
  });

  it('should have disabled create button if is_action_taken is true', () => {
    const { baseElement } = setup({
      ...MOCK_RETRO_ENTRY,
      is_action_taken: true
    });
    screen.debug();
    // get create button
    const createButton = screen.getByRole('button', { name: /create/i });

    // Assert that the button is disabled
    expect(createButton).toBeDisabled();
  })
});
