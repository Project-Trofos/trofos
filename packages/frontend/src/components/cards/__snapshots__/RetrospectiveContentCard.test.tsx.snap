// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RetrospectiveContentCard test > renders retrospective card with correct details 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered retrospective-content-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="retrospective-content-body"
        >
          <div
            class="vote-container"
          >
            <button
              class="ant-btn ant-btn-default up "
              type="button"
            >
              <span
                aria-label="arrow-up"
                class="anticon anticon-arrow-up"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="arrow-up"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M868 545.5L536.1 163a31.96 31.96 0 00-48.3 0L156 545.5a7.97 7.97 0 006 13.2h81c4.6 0 9-2 12.1-5.5L474 300.9V864c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V300.9l218.9 252.3c3 3.5 7.4 5.5 12.1 5.5h81c6.8 0 10.5-8 6-13.2z"
                  />
                </svg>
              </span>
            </button>
            <span
              class="vote-count"
            >
              1
            </span>
            <button
              class="ant-btn ant-btn-default down "
              type="button"
            >
              <span
                aria-label="arrow-down"
                class="anticon anticon-arrow-down"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="arrow-down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M862 465.3h-81c-4.6 0-9 2-12.1 5.5L550 723.1V160c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v563.1L255.1 470.8c-3-3.5-7.4-5.5-12.1-5.5h-81c-6.8 0-10.5 8.1-6 13.2L487.9 861a31.96 31.96 0 0048.3 0L868 478.5c4.5-5.2.8-13.2-6-13.2z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="retrospective-content-text"
          >
            Retrospective text
          </div>
          <div
            class="retrospective-content-delete-container"
          >
            <button
              class="ant-btn ant-btn-default delete"
              type="button"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
